### PR TITLE
feat: LemonModalV2

### PR DESCRIPTION
**Pull request by benjackwhite**
## Problem

Standardise all modals and remove usage of Antd Modal for our LemonModal in favor of much lighter-weight react-modal

## Changes

Different modals below:

Screenshot 2022-08-08 at 15 33 13
Screenshot 2022-08-08 at 15 33 22
Screenshot 2022-08-08 at 15 33 35
Screenshot 2022-08-08 at 15 52 56
Screenshot 2022-08-08 at 16 05 24

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## TODO

- [x] Fix Forms surrounding modals - SAML, AsyncMigrations
- [x] Fix modal in a modal (AddToDashboard - new dashboard)

### CODE REVIEW

**Review by clarkus — APPROVED** [review]
Noticed some small issues in the example modals, but I think those can be sorted out as new components are updated. The modal frame looks right to me. 👍

File: frontend/src/lib/components/Subscriptions/SubscriptionsModal.stories.tsx [hunk 1]
```diff
@@ -62,16 +62,19 @@ const Template = (
...
-            <div className="LemonModal">
-                <div className="rounded border ant-modal-body" style={{ width: 650, margin: '20px auto' }}>
-                    <Subscriptions
-                        {...(props as SubscriptionsModalProps)}
-                        closeModal={() => console.log('close')}
-                        insightShortId={insightShortIdRef.current}
-                        visible={true}
-                    />
-                </div>
+            {/* <div className="LemonModal"> */}
```
Comment (pauldambra) [inline frontend/src/lib/components/Subscriptions/SubscriptionsModal.stories.tsx:65]: Exorcise these comments ✝

File: frontend/src/lib/components/Subscriptions/SubscriptionsModal.tsx [hunk 2]
```diff
@@ -10,23 +9,25 @@ import { PayGatePage } from '../PayGatePage/PayGatePage'
...
+    const { closeModal, dashboardId, insightShortId, subscriptionId, visible, inline } = props
     const { push } = useActions(router)
     const { hasAvailableFeature, userLoading } = useValues(userLogic)
 
     if (userLoading) {
         return <Spinner />
     }
     return (
-        <>
+        <LemonModal onClose={closeModal} isOpen={visible} width={600} simple title="" inline={inline}>
```
Comment (pauldambra) [inline frontend/src/lib/components/Subscriptions/SubscriptionsModal.tsx:30]: `isOpen` nicer language than `visible` for a modal

but the repeated `isOpen={visible}` is yucky... rename the variables?

File: frontend/src/lib/constants.tsx [hunk 3]
```diff
@@ -116,7 +116,6 @@ export const FEATURE_FLAGS = {
     KAFKA_INSPECTOR: 'kafka-inspector', // owner: @yakkomajuri
     INSIGHT_EDITOR_PANELS: '8929-insight-editor-panels', // owner: @mariusandra
     FRONTEND_APPS: '9618-frontend-apps', // owner: @mariusandra
-    ONBOARDING_1_5: 'onboarding-1_5', // owner: @liyiy
```
Comment (pauldambra) [inline frontend/src/lib/constants.tsx:119]: this relies on @kappa90's remove onboarding PR?

File: frontend/src/scenes/organization/Settings/VerifiedDomains/AddDomainModal.tsx [hunk 4]
```diff
@@ -33,32 +33,32 @@ export function AddDomainModal(): JSX.Element {
...
-                    >
-                        Add domain
-                    </LemonButton>
-                </div>
-            </section>
+        <LemonModal
+            onClose={handleClose}
+            isOpen={addModalShown}
+            title="Add authentication domain"
+            footer={
```
Comment (pauldambra) [inline frontend/src/scenes/organization/Settings/VerifiedDomains/AddDomainModal.tsx:40]: We have some that take `footer` as a prop and some that have `Lemon.Footer` as a child...

Do we need both approaches? Does storybook teach us how to choose?

Comment (pauldambra) [reply frontend/src/scenes/organization/Settings/VerifiedDomains/AddDomainModal.tsx:40]: What happens if I add both :)

**Review by pauldambra — APPROVED** [review]
the `isOpen={visible}` is the only thing that was almost blocking for me. but easy to fix in follow-up :)

File: frontend/src/lib/components/LemonModal/LemonModal.scss [hunk 5]
```diff
@@ -1,57 +1,100 @@
...
+    box-shadow: 0 16px 16px -16px rgba(0, 0, 0, 0.35);
+
+    transition: opacity 250ms ease-out, transform 250ms ease-out;
+    display: flex;
+    flex-direction: col;
+    max-height: 90%;
+
+    // Transition properties
+    will-change: transform;
+    transform: scale(0.85) translateY(25%);
```
Comment (Twixes) [inline frontend/src/lib/components/LemonModal/LemonModal.scss:44]: The blur and scaling effects make for a nice experience, but I'm not so sure about the translate. Previously the translate effect originated where the click happened, and I don't think we need to keep that, but now the modal originates from a completely opposite place, which doesn't _feel_ ideal. I think it'd be nicer without `translate()`. Especially since the modal _disappears_ into the center in this implementation.

File: frontend/src/lib/components/LemonModal/LemonModal.scss [hunk 6]
```diff
@@ -1,57 +1,100 @@
...
+
+    display: flex;
+    align-items: center;
+    justify-content: center;
 
-    .ant-modal-header {
-        border-bottom: none;
+    &.ReactModal__Overlay--after-open {
+        background-color: rgba(0, 0, 0, 0.4);
+        backdrop-filter: blur(5px);
```
Comment (Twixes) [inline frontend/src/lib/components/LemonModal/LemonModal.scss:17]: I suggest putting this blur strength into a CSS variable, so that this is more explicitly part of the design system too (and sync with @lottiecoxon @clarkus so that designs include the backdrop blur too).

File: frontend/src/lib/components/LemonModal/LemonModal.scss [hunk 7]
```diff
@@ -1,57 +1,100 @@
...
-        border-color: var(--border);
-    }
+.LemonModal__overlay {
+    position: fixed;
+    top: 0;
+    left: 0;
+    right: 0;
+    bottom: 0;
+
+    transition: background-color 250ms ease-out, backdrop-filter 250ms ease-out;
```
Comment (Twixes) [inline frontend/src/lib/components/LemonModal/LemonModal.scss:8]: This feels _very slightly_ too long for a key interactive element. Why not keep it in sync with button transition times (200ms)? Better yet, our transition times should be standardizes as CSS vars

File: frontend/src/lib/components/LemonModal/LemonModal.tsx [hunk 8]
```diff
@@ -1,21 +1,99 @@
...
+    title,
+    description,
+    footer,
+    inline,
+    simple,
+}: LemonModalProps): JSX.Element {
+    const modalContent = (
+        <>
+            <div className="LemonModal__closebutton">
+                <LemonButton icon={<IconClose />} size="small" status="stealth" onClick={onClose} />
```
Comment (Twixes) [inline frontend/src/lib/components/LemonModal/LemonModal.tsx:55]: These workarounds with wrapping `div`s seem kind of pointless to me, this would be more straightforward if `LemonButton` had this as an extra `className` here.

Comment (Twixes) [inline frontend/src/lib/components/LemonModal/LemonModal.tsx:55]: Weirdly, this button isn't square, which looks a bit off as icons are square.

**Review by Twixes — CHANGES_REQUESTED** [review]
Really nice stuff, though large enough to warrant some suggestions. 😄 Also, what Paul said.

File: frontend/src/scenes/trends/PersonsModal.tsx [hunk 9]
```diff
@@ -167,144 +168,146 @@ export function PersonsModal({
...
-                                                            ? 'None'
-                                                            : dataPoint.breakdown_value?.toString()
-                                                    }
-                                                    showCountedByTag={showCountedByTag}
-                                                    hasMultipleSeries={hasMultipleSeries}
-                                                />
-                                            </Select.Option>
-                                        ))}
-                                    </Select>
+                <div className="persons-modal space-y-2">
```
Comment (Twixes) [inline frontend/src/scenes/trends/PersonsModal.tsx:172]: Side question: is it intentional that the list has a padding now?

| Before | After |
| --- | --- |
| Screen Shot 2022-08-09 at 12 34 45 | Screen Shot 2022-08-09 at 12 34 16 |

File: frontend/src/scenes/organization/Settings/VerifiedDomains/AddDomainModal.tsx (hunk 4 above)
Comment (benjackwhite) [reply frontend/src/scenes/organization/Settings/VerifiedDomains/AddDomainModal.tsx:40]: They are sadly necessary due to forms but for sure should be better documented. I'll update the storybook to be clear about it

File: frontend/src/lib/constants.tsx (hunk 3 above)
Comment (benjackwhite) [reply frontend/src/lib/constants.tsx:119]: Ah there might be some nasty stash stuff going on here. I'll clean it up

File: frontend/src/lib/components/Subscriptions/SubscriptionsModal.tsx (hunk 2 above)
Comment (benjackwhite) [reply frontend/src/lib/components/Subscriptions/SubscriptionsModal.tsx:30]: Yup was thinking the same so I should do it now otherwise I'll forget

File: frontend/src/lib/components/LemonModal/LemonModal.scss (hunk 5 above)
Comment (benjackwhite) [reply frontend/src/lib/components/LemonModal/LemonModal.scss:44]: Fair arguments. The animation isn't mentioned in the Design System so far. I thinkI could make it come from the direction of the button clicked with a _relatively_ small amount of code. Antd's implementation of this is pretty lightweight

File: frontend/src/lib/components/LemonModal/LemonModal.scss (hunk 6 above)
Comment (benjackwhite) [reply frontend/src/lib/components/LemonModal/LemonModal.scss:17]: Agreed. I abstracted it now and matched the Design System - 10px. To me it feels too strong but happy to defer to @clarkus / @lottiecoxon for input there

File: frontend/src/lib/components/LemonModal/LemonModal.scss (hunk 7 above)
Comment (benjackwhite) [reply frontend/src/lib/components/LemonModal/LemonModal.scss:8]: Read a lot of UX discussions on this, particularly Apple who argue that the bigger the element, the slower the transition as that gives it the right "weight" but definitely no opinion on what our values should be.

@clarkus side-discussion but perhaps Transitions should be covered in our Design System somewhere so that engineers can make good decisions here based on some sort of agreed basis principles?

File: frontend/src/lib/components/LemonModal/LemonModal.tsx (hunk 8 above)
Comment (benjackwhite) [reply frontend/src/lib/components/LemonModal/LemonModal.tsx:55]: Mentioned this in the other PR. I feel split on this. On one side I want to push for as little deviation from the design as possible hence limiting classnames, on the other hand it makes these useful helpers like layout harder...

Comment (benjackwhite) [reply frontend/src/lib/components/LemonModal/LemonModal.tsx:55]: I think this comes from the css line-height. I'll fix it in a follow up PR

File: frontend/src/scenes/trends/PersonsModal.tsx (hunk 9 above)
Comment (benjackwhite) [reply frontend/src/scenes/trends/PersonsModal.tsx:172]: Not intentional perse but intentionally not fixing it as this component deviates from the standard Modal styles anyways. This modal is incredibly buggy anyway so I am tempted to leave it this way and revisit with purpose.

File: frontend/src/lib/components/LemonModal/LemonModal.scss (hunk 5 above)
Comment (benjackwhite) [reply frontend/src/lib/components/LemonModal/LemonModal.scss]: @clarkus what do you think is the best compromise here?

File: frontend/src/lib/components/LemonModal/LemonModal.scss (hunk 6 above)
Comment (benjackwhite) [reply frontend/src/lib/components/LemonModal/LemonModal.scss]: Actually reverted to 5px as that matches the design closer. I think it's due to blur being based on resolution. Browsers cleverly scale up to "retina" so 5px is 10px with a higher DPI. _I think..._

|Figma (10px)|Code(5px)|Code(10px)|
|----|----|----|
|Screenshot 2022-08-09 at 17 00 01|Screenshot 2022-08-09 at 17 00 06|Screenshot 2022-08-09 at 17 00 59|

File: .all-contributorsrc [hunk 10]
```diff
@@ -1989,14 +1989,33 @@
...
+      ]
+    },
+    {
+      "login": "aviflombaum",
+      "name": "Avi Flombaum",
+      "avatar_url": "https://avatars.githubusercontent.com/u/4515?v=4",
+      "profile": "http://www.aviflombaum.com",
+      "contributions": [
+        "code"
+      ]
```
Comment (Twixes) [inline .all-contributorsrc:2018]: Bit odd this was added here 🤔

File: .all-contributorsrc [hunk 11]
```diff
@@ -1989,14 +1989,33 @@
...
     }
```
Comment (Twixes) [inline .all-contributorsrc:2019]: There's a bunch of unrelated changes like this in the diff, looks like this needs to have a different base

File: frontend/src/lib/components/LemonModal/LemonModal.scss (hunk 7 above)
Comment (clarkus) [reply frontend/src/lib/components/LemonModal/LemonModal.scss]: Yup we haven't worked on transitions much at all other than @lottiecoxon's animation work. It's something we can address for sure, but generally I think we should use transitions sparingly. While they can be a noticeable detail for some users, the end result is that things display more slowly. Too long of a transition can make your product feel clunky. Shorter transitions and a limited animation area are good goals. @lottiecoxon do you want to spend some time on transitions after you wrap up your current component work?

Comment (pauldambra) [reply frontend/src/lib/components/LemonModal/LemonModal.scss]: There's a media query you can use to disable animations if users have set reduced motion. Maybe something to consider as we think about how/when to use animations https://www.smashingmagazine.com/2020/09/design-reduced-motion-sensitivities/

Comment (clarkus) [reply frontend/src/lib/components/LemonModal/LemonModal.scss]: Good point, @pauldambra. Lottie and I were discussing the same thing yesterday regarding skeleton state animations. Something to consider with a design system are layers of configuration based on user preferences. The most popular options are dark mode and high contrast, but our current palette is high contrast by default. An option for limiting motion and animation would be a great thing to consider. Only mentioning color modes because they're often coupled with the reduced motion options for users.

File: .all-contributorsrc (hunk 11 above)
Comment (benjackwhite) [reply .all-contributorsrc:2019]: Yeah its dependent on the Search inputs work as I thought that would have been merged by now 😅

File: .all-contributorsrc (hunk 10 above)
Comment (benjackwhite) [reply .all-contributorsrc:2018]: See other comment

File: .all-contributorsrc (hunk 11 above)
Comment (benjackwhite) [reply .all-contributorsrc]: Updated to fix the weird diff.

**Review by Twixes — APPROVED** [review]
Looks great to me!
